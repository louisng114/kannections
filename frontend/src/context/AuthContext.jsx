import React, { createContext, useState, useEffect } from 'react';
import { jwtDecode } from "jwt-decode";
import KannectionsApi from '../helpers/api';

const AuthContext = createContext();

// functions for user authentication
const AuthProvider = ({ children }) => {
  const [user, setUser] = useState(null);

  useEffect(() => {
    const storedToken = localStorage.getItem('token');
    if (storedToken) {
      KannectionsApi.token = storedToken;
      fetchUserFromToken(storedToken);
    }
  }, []);

  const fetchUserFromToken = async (token) => {
    try {
      const decoded = jwtDecode(token);
      const user = await KannectionsApi.getUser(decoded.username);
      setUser(user);
    } catch (error) {
      console.error('Failed to decode token:', error);
    }
  };

  const signup = async (username, password) => {
    try {
        const token = await KannectionsApi.register({ username, password });
        localStorage.setItem('token', token);
        KannectionsApi.token = token;
        fetchUserFromToken(token);
    } catch (error) {
      console.error("Registration failed:", error);
      throw error;
    }
  };

  const login = async (username, password) => {
    try {
      const token = await KannectionsApi.getToken({ username, password });
      localStorage.setItem('token', token);
      KannectionsApi.token = token;
      fetchUserFromToken(token);
    } catch (error) {
      console.error("Login failed:", error);
      throw error;
    }
  };

  const logout = () => {
    localStorage.removeItem('token');
    setUser(null);
    KannectionsApi.token = null;
  };

  return (
    <AuthContext.Provider value={{ user, signup, login, logout }}>
      {children}
    </AuthContext.Provider>
  );
};

export { AuthProvider, AuthContext };
