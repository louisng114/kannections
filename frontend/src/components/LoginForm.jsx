import React, { useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import { useFormik } from 'formik';
import { AuthContext } from '../context/AuthContext';
import * as Yup from 'yup';
import { Form, Button, Container } from 'react-bootstrap';
import "./Form.css";

const LoginForm = () => {
    const navigate = useNavigate();
    const { login } = useContext(AuthContext);

    const formik = useFormik({
        initialValues: {
            username: '',
            password: '',
        },
        validationSchema: Yup.object({
            username: Yup.string().required('Required'),
            password: Yup.string().min(5, 'Password must be at least 5 characters').required('Required'),
        }),
        onSubmit: async (values) => {
            try {
                await login(values.username, values.password);
                navigate('/');
            } catch (error) {
                formik.setFieldError('general', 'Invalid credentials');
            }
        },
    });

    return (
        <Container className="mt-5">
            <Form onSubmit={formik.handleSubmit}>
                <Form.Group controlId="username">
                    <Form.Label>Username</Form.Label>
                    <Form.Control
                        type="text"
                        name="username"
                        onChange={formik.handleChange}
                        onBlur={formik.handleBlur}
                        value={formik.values.username}
                        isInvalid={formik.touched.username && !!formik.errors.username}
                        autoComplete="username"
                    />
                    <Form.Control.Feedback type="invalid">
                        {formik.errors.username}
                    </Form.Control.Feedback>
                </Form.Group>
                
                <Form.Group controlId="password">
                    <Form.Label>Password</Form.Label>
                    <Form.Control
                        type="password"
                        name="password"
                        onChange={formik.handleChange}
                        onBlur={formik.handleBlur}
                        value={formik.values.password}
                        isInvalid={formik.touched.password && !!formik.errors.password}
                        autoComplete="current-password"
                    />
                    <Form.Control.Feedback type="invalid">
                        {formik.errors.password}
                    </Form.Control.Feedback>
                </Form.Group>
                
                {formik.errors.general && (
                    <div className="text-danger mb-3">{formik.errors.general}</div>
                )}
                <div className='form-button'>
                    <Button type="submit" disabled={formik.isSubmitting}>
                        Log In
                    </Button>
                </div>
            </Form>
        </Container>
    );
};

export default LoginForm;
