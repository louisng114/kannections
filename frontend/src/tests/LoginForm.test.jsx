import { render } from '@testing-library/react';
import LoginForm from '../components/LoginForm';
import { AuthProvider } from '../context/AuthContext';
import { BrowserRouter } from 'react-router-dom';

test('renders without crashing', () => {
  render(
    <AuthProvider>
      <BrowserRouter>
        <LoginForm />
      </BrowserRouter>
    </AuthProvider>
  );
});

it("matches snapshot", function() {
  const {asFragment} = render(
      <AuthProvider>
        <BrowserRouter>
          <LoginForm />
        </BrowserRouter>
      </AuthProvider>
    );
  expect(asFragment()).toMatchSnapshot();
});
