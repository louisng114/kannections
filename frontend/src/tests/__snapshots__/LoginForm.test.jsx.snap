// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="mt-5 container"
  >
    <form
      class=""
    >
      <div>
        <label
          class="form-label"
          for="username"
        >
          Username
        </label>
        <input
          autocomplete="username"
          class="form-control"
          id="username"
          name="username"
          type="text"
          value=""
        />
        <div
          class="invalid-feedback"
        />
      </div>
      <div>
        <label
          class="form-label"
          for="password"
        >
          Password
        </label>
        <input
          autocomplete="current-password"
          class="form-control"
          id="password"
          name="password"
          type="password"
          value=""
        />
        <div
          class="invalid-feedback"
        />
      </div>
      <div
        class="form-button"
      >
        <button
          class="btn btn-primary"
          type="submit"
        >
          Log In
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
